### ISSUE TITLE
Squid does not always log the right source IP

### ISSUE THREAD

**Issue by yourname43** [e2guardian/e2guardian#212]
Sometimes the source IP is right, sometimes it's wrong :

* https :
```
root@amon:~# tail -f log/rsyslog/local/squid/squid1.info.log|grep www.admission-postbac.fr
2017-04-14T10:00:28.807113+02:00 amon.etb1.lan (squid-1): 1492156828.806      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:28.828043+02:00 amon.etb1.lan (squid-1): 1492156828.827      1 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.053131+02:00 amon.etb1.lan (squid-1): 1492156831.052      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.054492+02:00 amon.etb1.lan (squid-1): 1492156831.053      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.060861+02:00 amon.etb1.lan (squid-1): 1492156831.060      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.061559+02:00 amon.etb1.lan (squid-1): 1492156831.061      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.062133+02:00 amon.etb1.lan (squid-1): 1492156831.062      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.082375+02:00 amon.etb1.lan (squid-1): 1492156831.082      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.083899+02:00 amon.etb1.lan (squid-1): 1492156831.083      1 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.130871+02:00 amon.etb1.lan (squid-1): 1492156831.130      2 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.131374+02:00 amon.etb1.lan (squid-1): 1492156831.130      2 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.131647+02:00 amon.etb1.lan (squid-1): 1492156831.130      0 10.1.2.51 TCP_DENIED/407 4258 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.132326+02:00 amon.etb1.lan (squid-1): 1492156831.132      2 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:31.147646+02:00 amon.etb1.lan (squid-1): 1492156831.147      1 10.1.2.51 TCP_DENIED/407 4557 CONNECT www.admission-postbac.fr:443 - HIER_NONE/- text/html
2017-04-14T10:00:36.313386+02:00 amon.etb1.lan (squid-1): 1492156836.312   5149 127.0.0.1 TCP_TUNNEL/200 9375 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -
2017-04-14T10:00:36.314100+02:00 amon.etb1.lan (squid-1): 1492156836.312   5133 127.0.0.1 TCP_TUNNEL/200 6101 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -
2017-04-14T10:00:36.314306+02:00 amon.etb1.lan (squid-1): 1492156836.313   5122 127.0.0.1 TCP_TUNNEL/200 4345 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -
2017-04-14T10:00:36.314470+02:00 amon.etb1.lan (squid-1): 1492156836.313   7446 127.0.0.1 TCP_TUNNEL/200 37020 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -
2017-04-14T10:00:36.314658+02:00 amon.etb1.lan (squid-1): 1492156836.313   5129 127.0.0.1 TCP_TUNNEL/200 6257 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -
2017-04-14T10:00:36.441405+02:00 amon.etb1.lan (squid-1): 1492156836.440   5257 127.0.0.1 TCP_TUNNEL/200 12137 CONNECT www.admission-postbac.fr:443 admin HIER_DIRECT/193.50.43.66 -

```

* http :

```
2017-04-14T10:08:59.034313+02:00 amon.etb1.lan (squid-1): 1492157339.034      0 10.1.2.51 TCP_DENIED/407 4366 GET http://bp-eole.ac-dijon.fr/ - HIER_NONE/- text/html
2017-04-14T10:09:04.145227+02:00 amon.etb1.lan (squid-1): 1492157344.145      1 10.1.2.51 TCP_DENIED/407 4665 GET http://bp-eole.ac-dijon.fr/ - HIER_NONE/- text/html
2017-04-14T10:09:04.199843+02:00 amon.etb1.lan (squid-1): 1492157344.199     12 127.0.0.1 TCP_MISS/200 526 GET http://bp-eole.ac-dijon.fr/ admin HIER_DIRECT/194.167.18.21 text/html
2017-04-14T10:09:04.339816+02:00 amon.etb1.lan (squid-1): 1492157344.339      3 10.1.2.51 TCP_MISS/404 617 GET http://bp-eole.ac-dijon.fr/favicon.ico admin HIER_DIRECT/194.167.18.21 text/html
```

**Comment by yourname43** [comment]
Here is a wireshark capture 
[e2g-4.1_127.0.0.1err.zip](https://github.com/e2guardian/e2guardian/files/921930/e2g-4.1_127.0.0.1err.zip)

**Comment by fredbcode** [comment]
Line 32 http.x_forwarded_for is not present and line 24 it is twice 

Line 32:
Hypertext Transfer Protocol
    CONNECT www.admission-postbac.fr:443 HTTP/1.0\r\n
    User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:52.0) Gecko/20100101 Firefox/52.0\r\n
    Connection: close\r\n
    Connection: keep-alive\r\n
    Host: www.admission-postbac.fr\r\n
     [truncated]Proxy-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAH4AAAAcARwBlgAAAAgACABYAAAACgAKAGAAAAAUABQAagAAAAAAAACyAQAABYKIogoAOTgAAAAPpqMMzoUpdQCyzdYSb0HvS0UAVABCADEAYQBkAG0AaQBuAEIATABBAEIATABBAFMARQBUAEgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA0
    \r\n
    [Full request URI: http://www.admission-postbac.frwww.admission-postbac.fr:443]
    [HTTP request 2/2]
    [Prev request in frame: 24]
    [Response in frame: 34]

Line 24:

Hypertext Transfer Protocol
    CONNECT www.admission-postbac.fr:443 HTTP/1.0\r\n
    User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:52.0) Gecko/20100101 Firefox/52.0\r\n
    Connection: keep-alive\r\n
    Connection: keep-alive\r\n
    Host: www.admission-postbac.fr\r\n
    Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB4IIogAAAAAAAAAAAAAAAAAAAAAKADk4AAAADw==\r\n
    X-Forwarded-For: 10.1.2.51\r\n
    X-Forwarded-For: 10.1.2.51\r\n
    \r\n
    [Full request URI: http://www.admission-postbac.frwww.admission-postbac.fr:443]
    [HTTP request 1/2]
    [Response in frame: 28]
    [Next request in frame: 32]

A correct value 

Hypertext Transfer Protocol
    CONNECT www.admission-postbac.fr:443 HTTP/1.0\r\n
    User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:52.0) Gecko/20100101 Firefox/52.0\r\n
    Connection: keep-alive\r\n
    Connection: keep-alive\r\n
    Host: www.admission-postbac.fr\r\n
    X-Forwarded-For: 10.1.2.51\r\n
    \r\n
    [Full request URI: http://www.admission-postbac.frwww.admission-postbac.fr:443]
    [HTTP request 1/1]
    [Response in frame: 166]

There is certainly a bug here 

Which version are you using ?

**Comment by fredbcode** [comment]
Also twice "Connection: keep-alive\r\n" ??

**Comment by Kenny690** [comment]
I have the same issue. Sometimes squid writes it's own IP except of writing clients IP. I have NTLM so sometime it asking user for a password in that moments. When I'm moving squid to port 8081 - everything's getting to be ok.

**Comment by Kenny690** [comment]
Hm... Found something new. I tried to set proxyexchange = 5 and I've got mail.ru in 5 sec. Maybe the problem is in proxy. I will investigate it.

**Comment by fredbcode** [comment]
Some changes in latest push (v4.1)

Tested with squid 3.5.12 on Ubuntu 16.04

I added in squid.conf: **follow_x_forwarded_for allow all**

Tested with https://www.google.fr, https://youtube.com and http://numsys.eu
I can see the original addresses with NTLM

**Comment by fredbcode** [comment]
Please, could someone do a test ?

**Comment by Kenny690** [comment]
This problem's seems like gone.

**Comment by fredbcode** [comment]
Ok great thanks for your help

**Comment by fredbcode** [comment]
Apr 26 12:15:40 osiris e2guardian[10600]: NTLM - Invalid message of length 0, message was:
Apr 26 12:15:40 osiris e2guardian[10600]: Auth plugin returned error code: -3

Xforwarded should be added with no NTLM requests

**Comment by fredbcode** [comment]
This messages should be removed now

**Comment by Kenny690** [comment]
Nope. It's still there.

**Comment by fredbcode** [comment]
Weird not one here, except one or two at start, because NTLM session are broken, but after that nothing

**Comment by fredbcode** [comment]
Ah maybe which browser ? I'm using Firefox

**Comment by Kenny690** [comment]
Chrome. We all using chrome.

**Comment by fredbcode** [comment]
Ok I pushed many changes with NTLM, tested with IE, Chrome and FF

**Comment by Kenny690** [comment]
It works! Except one thing, but maybe it's normal. If user has no NTLM auth, then sslmitm is turned off. 
Example (the same IP, but no mitm without auth):
```
2017.4.27 13:38:14 kossa 10.20.30.138 https://rs.mail.ru/d25615525.gif?&rnd=148454245&test_id=336 - GET 43 0 - 1 200 image/gif - default - - - - -
2017.4.27 13:38:14 - 10.20.30.138 https://rs.mail.ru:443 - CONNECT 3674 0 - 1 0 - - default - - - - -

```

**Comment by fredbcode** [comment]
Ok good, I'm trying to reproduce, so I removed all authpluing (no identification at all) and SSLMITM works

2017.4.27 10:49:16 - 10.x.x.x.x https://live.github.com/_sockets/VjI6MTE5MzA3NzA4OmI5dmJmNWQ1NWU5NTM5NjlhZjMxODNiZmVjZTM1N2I2MDcyZGRmNjY0MzU5NTMzODFmMGNiYzI5Yjc4YTRdNDY=--476bd4bf71a6fd7f118fed7181a85b6397d78915 - GET 0 0 - 1 200 text/event-stream - default - - - - -

Please, can you make a try like me ?

**Comment by fredbcode** [comment]
> Example (the same IP, the same website, but no mitm without auth):

How you do that ? At the same time a request with and without auth

**Comment by Kenny690** [comment]
I turned off auth plugins. SSLMITM works fine without that. But after I turned on sslmitm again not working with unauthed users. 
BTW, I found something new. I turned on, for test, IP-based auth before NTLM (so there's no NTLM auth). The same issue. No sslmitm. 

`2017.4.27 14:20:42 10.20.30.237 10.20.30.237 https://googleads.g.doubleclick.net:443 - CONNECT 3706 0 - 1 0 - - default - - - - -
`

Then I put IP after NTLM. When there is IP auth - no sslmitm, but NTLM users has it.

```
2017.4.27 14:26:17 10.20.30.114 10.20.30.114 https://push.yandex.ru:443 - CONNECT 3680 0 - 1 0 - - default - - - - -
2017.4.27 14:26:22 karma 10.20.30.114 https://push.yandex.ru/v1/subscribe?client=morda&service=mail&session=1493283208-98805-20945-21360&uid=73257651 - GET 0 0 - 1 101 - - default - - - - -
```

**Comment by fredbcode** [comment]
Can you make a try with IP-Based only ?

**Comment by philipianpearce** [comment]
I know this may seem a silly question, but do you have sslmitm =on in e2guardianf1.conf ? 

Regards 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
Philip I can reproduce, I certainly made a mistake somewhere 
With ntlm all seems right, IP, username, etc but no SSLMITM, if "off" it works 

I'm trying to understand what could proprely disabled SSLMITM ! is_ssl ?

**Comment by Kenny690** [comment]
Sure I have it turned on. It could be an another bug if it was turned off, cause why do I ever have sslmitm on NTLM. :D
With IP-Based only I have sslmitm working fine.

**Comment by fredbcode** [comment]
Arg ! sslmi**n**tm = on in the group file, fixed now 

Sorry now it works for me here NTLM + SSLMITM 

I guess I understand better the problem now 

Some "wrong" NTLM requests now pass and are checked as group1 

Like this:
2017.4.27 13:38:14 - 10.20.30.138 https://rs.mail.ru:443 - CONNECT 3674 0 - 1 0 - - default - - - - -

But 
2017.4.27 13:38:14 kossa 10.20.30.138 https://rs.mail.ru/d25615525.gif?&rnd=148454245&test_id=336 - GET 43 0 - 1 200 image/gif - default - - - - -

Do you have sslmitm = on on all your e2guardianfx.conf (with the right syntax :) )

**Comment by Kenny690** [comment]
Yes. Actually, after every upgrade of binary - I'm restoring my configs from the backup. :D

**Comment by fredbcode** [comment]
I think I know where the problem is, If you compare squid log and e2guardian you see the 407 requests are loggued as group1 in e2guardian 

And with NTLM there are a lot of 407 ... 

It's what 0 means here 
2017.4.27 13:38:14 - 10.20.30.138 https://rs.mail.ru:443 - CONNECT 3674 0 - 1 **0** - - default 

Can you confirm Kenny ?

**Comment by Kenny690** [comment]
Yes. 
![default](https://cloud.githubusercontent.com/assets/4517038/25479775/7fd416a4-2b5f-11e7-88d5-0f1689df4ddb.PNG)

**Comment by fredbcode** [comment]
In negotiation with some IE and Chrome only the header is corrupted 
The result of h.in(&peercon, true) is NULL in this case (and only) the request is allowed

We must remake the identification part, at least for NTLM ...

Until this I have a decision to make:

1 - I remove this requests from log, to avoid any confusion (I push a fix like this for testing)
2 - Let this, and we must known that this kind of log is a part of negotiation ... 

Before we had also removed redirection code from logs, so logs full is not totally right ...

@philipianpearce What do you think ?
@Kenny690 can you make a test ?

**Comment by Kenny690** [comment]
All the same.

**Comment by fredbcode** [comment]
E2guardian now show the right code "407" I was totally lost in the spaghetti code 
There are 12 call to doLog ...

I guess this bug was not related with NTLM, maybe NTLM+SSLMITM 

@Kenny690 thanks for your help now you should see:

2017.4.27 13:38:14 - 10.20.30.138 https://rs.mail.ru:443 - CONNECT 3674 0 - 1 0 **407**- default 
exactly like Squid

The 407 requests are a part of NTLM negotiation, don't worry about 

@philipianpearce I think there is a more important bug here, in ConnectionHandler.cpp before line 2570 docheader.returnCode() is always empty, I don't know exactly where and why but DoLog is always wrong before this limit.
I don't know it's this issue is related with SSLMITM or/and NTLM, but today I'm fed with it :)

**Comment by philipianpearce** [comment]
@fredb re docheader.returnCode() issues - this is what I was fixing in 'header' branch - it will always be 0 until the first time docheader.in() is called (line 1360) - when a non-http first header is returned from server it was doing a segv - with 'header' this should not happen as HTTPHeader.in() is now checking for valid 1st line header. 

I do not understand why we are now logging 407's as this has never happened before?? I'll have a look at the changes you have made next week. 

Regards 

Philip 

----- Original Message -----

**Comment by fredbcode** [comment]
> 
> I do not understand why we are now logging 407's as this has never
> happened before?? I'll have a look at the changes you have made next
> week.

With SSLMITM enabled all the 407 lines are logged here, but like the value is empty we have many wrong lines
Perhaps we should remove from process all headers with 407, 301, 302, etc, (before filtering and identification) it should help a lot the engine, especially with SSLMITM
I mean add to exception all this HTTP codes and not decrypt, ident, filtering this pages 

I put the 407 to see if I'm right, now all works good but I agree this need a proper code, it's just a proof of concept 
Maybe this is only related with NTLM, I don't know 
I spend much time with this, but I will make tests with BASIC or DIGEST + SSLMITM soon

With some browsers (IE and Chrome) h.header.size is also empty in negotiation part (and only sometime ...) so I added an exception.
https://github.com/e2guardian/e2guardian/blob/v4.1/src/authplugins/ntlm.cpp#L266

Great minds think alike :) I will take a look a "header"

Fred

**Comment by fredbcode** [comment]
Ok now I fixed this problem with a proper fix

Usual (200, 304, etc) : logged
Auth (407) : logged (NTLM/SSLMITM or not)
Redirection (302, etc) : logged
Exception (I mean unexpected value) : logged

 > this is what I was fixing in 'header' branch - it will always be 0 until the first time docheader.in() is called (line 1360)

When it's done what do you think of move this block https://github.com/e2guardian/e2guardian/blob/v4.1/src/ConnectionHandler.cpp#L2577 before ?

At soon as possible we should exclude many codes from filtering or SSLMITM, It should reduce cycle usage (at least with NTLM 3x 407 for each request) ?
In this latest version I only let 200 and 304 but perhaps I forgot something

In log we should also add a special group tag for 407 now the request is in group1 because the user is unauthenticated

**Comment by fredbcode** [comment]
Ok fixed:
Group value = "negociate_identification"

2017.4.28 15:47:29 - 10.253.x.x http://detectportal.firefox.com/success.txt - GET 56 0 - 1 407 - - negociate_identification - - - - -
2017.4.28 15:47:29 - 10.253.x.x http://detectportal.firefox.com/success.txt - GET 56 0 - 1 407 - - negociate_identification - - - - -
2017.4.28 15:47:31 fbourgeois 10.253.x.x http://detectportal.firefox.com/success.txt - GET 8 0 - 1 200 text/plain - group1 - - - - -
2017.4.28 15:47:31 fbourgeois 10.253.x.x http://detectportal.firefox.com/success.txt - GET 8 0 - 1 200 text/plain - group1 - - - - -

**Comment by fredbcode** [comment]
Good for you now ?

**Comment by fredbcode** [comment]
Fixed in the next 4.1.1

**Comment by fredbcode** [comment]
There are two bugs here the first (specific to NTLM) is now fixed  
Still a problem, eg: if I disabled all plugins in E2 but I'm still using an identification method with Squid (Like IP acl) some requests are logged with 127.0.0.1
